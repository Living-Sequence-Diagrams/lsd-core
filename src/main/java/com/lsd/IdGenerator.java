package com.lsd;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

import static java.util.UUID.randomUUID;

public class IdGenerator {
    private final AtomicInteger counter = new AtomicInteger();

    private final Supplier<String> idSupplier;

    public IdGenerator(boolean isDeterministic) {
        if (isDeterministic) {
            idSupplier = () -> String.valueOf(counter.incrementAndGet());
        } else {
            idSupplier = () -> randomUUID().toString()
                    .replaceAll("-", "");
        }
    }

    public String next() {
        return idSupplier.get();
    }

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    public void reset() {
        counter.set(0);
    }
}
