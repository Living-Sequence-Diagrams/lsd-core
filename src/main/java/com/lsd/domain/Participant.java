package com.lsd.domain;

import lombok.Value;

@Value
public class Participant {
    String markup;
}
