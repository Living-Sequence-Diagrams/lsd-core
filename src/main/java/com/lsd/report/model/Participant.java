package com.lsd.report.model;

import lombok.Value;

@Value
public class Participant {
    String markup;
}
