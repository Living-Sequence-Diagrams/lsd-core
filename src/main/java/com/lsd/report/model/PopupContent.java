package com.lsd.report.model;

import j2html.tags.ContainerTag;
import j2html.tags.UnescapedText;

import static j2html.TagCreator.*;

/**
 * Simplifies adding content that is only displayed when a hyperlink is clicked - the id used on the href must match.
 */
public class PopupContent {

    public static ContainerTag popupDiv(String id, String title, String content) {
        return div().withId(id).withClass("overlay").attr("onclick", "location.href='#!';").with(
                div().withClass("popup").attr("onclick", "event.stopPropagation();").with(
                        h2(title),
                        a().withClass("close").withHref("#!").with(new UnescapedText("&times;")),
                        div().withClass("content").with(
                                pre(content)
                        )
                )
        );
    }
}
