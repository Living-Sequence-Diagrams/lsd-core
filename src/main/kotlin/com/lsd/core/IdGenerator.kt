package com.lsd.core

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger

private val randomIdSupplier: () -> String = { randomUUID().toString().replace("-".toRegex(), "") }

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    private val deterministicIdSupplier: () -> String = { counter.incrementAndGet().toString() }

    private val idSupplier: () -> String = if (isDeterministic) deterministicIdSupplier else randomIdSupplier

    fun next(): String = idSupplier.invoke()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}