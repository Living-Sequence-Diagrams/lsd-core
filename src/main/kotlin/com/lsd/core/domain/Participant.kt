package com.lsd.core.domain

data class Participant(
    var type: ParticipantType,
    val component: ComponentName,
    val alias: String?
)

enum class ParticipantType {
    ACTOR,
    BOUNDARY,
    COLLECTIONS,
    CONTROL,
    DATABASE,
    ENTITY,
    PARTICIPANT,
    QUEUE;

    fun called(name: String) = called(name = name, alias = null)

    fun called(name: String, alias: String?) =
        Participant(component = ComponentName(name), type = this, alias = alias)
}