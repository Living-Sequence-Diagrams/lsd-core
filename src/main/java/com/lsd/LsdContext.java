package com.lsd;

import com.lsd.diagram.ComponentDiagramGenerator;
import com.lsd.diagram.SequenceDiagramGenerator;
import com.lsd.events.SequenceEvent;
import com.lsd.parse.MessageParser;
import com.lsd.parse.Parser;
import com.lsd.parse.SynchronousResponseParser;
import com.lsd.properties.LsdProperties;
import com.lsd.report.HtmlIndexWriter;
import com.lsd.report.HtmlReportRenderer;
import com.lsd.report.HtmlReportWriter;
import com.lsd.report.model.DataHolder;
import com.lsd.report.model.Participant;
import com.lsd.report.model.Report;
import com.lsd.report.model.Scenario;

import java.nio.file.Path;
import java.util.*;

import static com.lsd.properties.LsdProperties.DETERMINISTIC_IDS;
import static com.lsd.properties.LsdProperties.MAX_EVENTS_PER_DIAGRAM;
import static java.util.stream.Collectors.toList;

public class LsdContext {

    private static final LsdContext INSTANCE = new LsdContext();

    private final List<CapturedScenario> capturedScenarios = new ArrayList<>();
    private final List<CapturedReport> capturedReports = new ArrayList<>();
    private final List<Participant> participants = new ArrayList<>();
    private final Set<String> includes = new LinkedHashSet<>();
    private final IdGenerator idGenerator = new IdGenerator(LsdProperties.getBoolean(DETERMINISTIC_IDS));
    private final List<Parser> parsers = parsers();
    private final HtmlReportWriter htmlReportWriter = new HtmlReportWriter(new HtmlReportRenderer());
    private final int maxEventsPerDiagram = LsdProperties.getInt(MAX_EVENTS_PER_DIAGRAM);

    private CapturedScenario currentScenario = new CapturedScenario();

    /**
     * Convenience method to create a singleton instance.
     * If you prefer to use other means to share a single instance or
     * want to create multiple instances then use the default constructor instead.
     *
     * @return A singleton instance of LsdContext
     */
    public static LsdContext getInstance() {
        return INSTANCE;
    }

    public void addParticipants(List<Participant> additionalParticipants) {
        participants.addAll(additionalParticipants);
    }

    public void includeFiles(Set<String> additionalIncludes) {
        includes.addAll(additionalIncludes);
    }

    public void addFact(String key, String value) {
        currentScenario.addFact(key, value);
    }

    public void capture(SequenceEvent event) {
        currentScenario.add(event);
    }

    /**
     * Allow string representations of events to be interpreted. If none of the parsers match the pattern then nothing will be captured
     *
     * @param pattern The input string that represents some event to be captured on the sequence diagram
     * @param body    The extra data associated with the event (may be null).
     */
    public void capture(String pattern, String body) {
        var event = parsers.stream()
                .map(parser -> parser.parse(pattern, body))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .findFirst();

        event.ifPresent(this::capture);
    }

    public void completeScenario(String title, String description, OutcomeStatus status) {
        currentScenario.setTitle(title);
        currentScenario.setDescription(description);
        currentScenario.setOutcomeStatus(status);
        capturedScenarios.add(currentScenario);
        currentScenario = new CapturedScenario();
    }

    public Path completeReport(String title) {
        Report report = buildReport(title);
        Path path = htmlReportWriter.writeToFile(report);
        capturedReports.add(new CapturedReport(report.getTitle(), path.getFileName().toString(), report.getStatus()));
        capturedScenarios.clear();
        currentScenario = new CapturedScenario();
        return path;
    }

    public String generateReport(String title) {
        Report report = buildReport(title);
        return htmlReportWriter.writeToString(report);
    }

    public Path createIndex() {
        return HtmlIndexWriter.writeToFile(capturedReports);
    }

    public void clear() {
        idGenerator.reset();
        capturedScenarios.clear();
        capturedReports.clear();
        currentScenario = new CapturedScenario();
    }

    public IdGenerator getIdGenerator() {
        return idGenerator;
    }

    private Report buildReport(String title) {
        return Report.builder()
                .title(title)
                .status(determineOverallStatus(capturedScenarios))
                .scenarios(capturedScenarios.stream()
                        .map(capturedScenario -> Scenario.builder()
                                .title(capturedScenario.getTitle())
                                .id(idGenerator.next())
                                .status(capturedScenario.getOutcomeStatus().getCssClass())
                                .description(capturedScenario.getDescription())
                                .facts(capturedScenario.getFacts())
                                .dataHolders(capturedScenario.getSequenceEvents().stream()
                                        .filter(DataHolder.class::isInstance)
                                        .map(DataHolder.class::cast)
                                        .collect(toList()))
                                .sequenceDiagram(SequenceDiagramGenerator.builder()
                                        .idGenerator(idGenerator)
                                        .events(capturedScenario.getSequenceEvents())
                                        .participants(participants)
                                        .includes(includes)
                                        .build()
                                        .diagram(maxEventsPerDiagram)
                                        .orElse(null))
                                .componentDiagram(ComponentDiagramGenerator.builder()
                                        .idGenerator(idGenerator)
                                        .events(capturedScenario.getSequenceEvents())
                                        .participants(participants)
                                        .build()
                                        .diagram()
                                        .orElse(null))
                                .build())
                        .collect(toList()))
                .build();
    }

    private String determineOverallStatus(List<CapturedScenario> capturedScenarios) {
        return capturedScenarios.stream()
                .map(CapturedScenario::getOutcomeStatus)
                .sorted()
                .findFirst()
                .map(OutcomeStatus::getCssClass)
                .orElse("");
    }

    private List<Parser> parsers() {
        return List.of(
                new MessageParser(idGenerator),
                new SynchronousResponseParser(idGenerator)
        );
    }
}
