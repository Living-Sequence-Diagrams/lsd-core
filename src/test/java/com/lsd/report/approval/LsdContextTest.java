package com.lsd.report.approval;

import com.lsd.IdGenerator;
import com.lsd.LsdContext;
import com.lsd.events.*;
import com.lsd.report.model.Participant;
import com.lsd.report.model.PopupContent;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.StringWriter;
import java.io.Writer;
import java.util.LinkedHashSet;
import java.util.List;

import static com.lsd.OutcomeStatus.*;
import static com.lsd.ParticipantType.*;
import static com.lsd.events.ArrowType.*;
import static j2html.TagCreator.*;

class LsdContextTest {

    private final LsdContext lsdContext = LsdContext.getInstance();
    private final IdGenerator idGenerator = lsdContext.getIdGenerator();

    private LinkedHashSet<String> additionalIncludes = new LinkedHashSet<>(List.of(
            "tupadr3/font-awesome-5/hamburger",
            "tupadr3/font-awesome-5/heart"
    ));

    private final List<Participant> participants = List.of(
            ACTOR.called("A", "Arnie"),
            BOUNDARY.called("Unused participant"),
            DATABASE.called("B", "Badboy\\nDB")
    );

    @BeforeEach
    public void clearContext() {
        lsdContext.clear();
    }

    @Test
    void createsReportWithScenariosAndEvents() {
        lsdContext.addParticipants(participants);
        lsdContext.includeFiles(additionalIncludes);

        lsdContext.capture(ShortMessageInbound.builder().id(nextId()).to("A").label("in").data("start some job").build());
        lsdContext.capture(Message.builder().id(nextId()).from("A").to("B").label("Message 1").data("some data 1").arrowType(BI_DIRECTIONAL).build());
        lsdContext.capture(ShortMessageOutbound.builder().id(nextId()).from("B").label("out").data("some data 1").build());
        lsdContext.capture(Message.builder().id(nextId()).label("An interaction description that is long enough to need abbreviating").from("Beta").to("Gamma").data("Î²").arrowType(LOST).build());
        lsdContext.capture(SynchronousResponse.builder().id(nextId()).label("A synchronous response").from("Gamma").to("Beta").data("200 OK").build());
        lsdContext.completeScenario("A Success scenario", "First scenario description", SUCCESS);

        lsdContext.capture(Message.builder().id(nextId()).label("Sending food <$hamburger{scale=0.4}>").from("A").to("B").colour("orange").arrowType(DOTTED_THIN).build());
        lsdContext.capture(new Markup("..."));
        lsdContext.capture("Sending a response from B to A [#red]", "Thank You!");
        lsdContext.capture(new NoteLeft("Friends <$heart{scale=0.4,color=red}>"));
        lsdContext.addFact("Something to highlight", "Lorem");
        lsdContext.addFact("Something else to highlight", "amet");
        lsdContext.addFact("Something else to highlight", "consectetur");
        lsdContext.addFact("Something else to highlight", "Thank you!");
        lsdContext.completeScenario("A Warning scenario", p(
                text("A popup with a long text that needs scrolling: "),
                a().withHref("#" + "kljasdlfj").withText("click me!"),
                PopupContent.popupDiv("kljasdlfj", "I am popup", ".. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.")
        ).render(), WARN);

        lsdContext.capture("a request from Beta to Gamma", "Please do something");
        lsdContext.capture("sync a synchronous response from Gamma to Beta", "Some Error (123456)");
        lsdContext.addFact("some important value", "123456");
        lsdContext.completeScenario("An Error scenario", "<p>Failure! Expected value to be 123 but was 123456</p>", ERROR);

        Approvals.verify(lsdContext.completeReport("Approval Report").toFile());
    }

    @Test
    void hasNoDiagramSectionWhenThereAreNoEvents() {
        lsdContext.addParticipants(participants);
        lsdContext.includeFiles(additionalIncludes);
        lsdContext.addFact("Fact", "no diagrams");
        lsdContext.completeScenario("A scenario without events", "This has no diagrams", SUCCESS);
        
        Approvals.verify(lsdContext.completeReport("No Diagrams Report").toFile());
    }

    @Test
    void createsIndex() {
        lsdContext.completeScenario("Scenario 1", "Error", ERROR);
        lsdContext.completeScenario("Scenario 2", "Warn", WARN);
        lsdContext.completeScenario("Scenario 3", "Success", SUCCESS);
        lsdContext.completeReport("First Report (Error)");

        lsdContext.completeScenario("Scenario 1", "Warn", WARN);
        lsdContext.completeScenario("Scenario 2", "Success", SUCCESS);
        lsdContext.completeReport("Second Report (Warn)");

        lsdContext.completeScenario("Scenario 1", "Success", SUCCESS);
        lsdContext.completeReport("Third Report");

        lsdContext.completeReport("Fourth Report");

        Approvals.verify(lsdContext.createIndex().toFile());
    }

    @Test
    void generateReport() {
        Writer writer = new StringWriter();

        String result = lsdContext.generateReport("Some title", writer);

        Approvals.verify(result);
    }
    private String nextId() {
        return idGenerator.next();
    }
}