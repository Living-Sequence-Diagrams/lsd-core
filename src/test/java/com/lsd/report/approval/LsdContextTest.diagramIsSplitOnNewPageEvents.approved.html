<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns="http://www.w3.org/1999/html" xml:lang="en" lang="en">
<head>
    <title>Split by NewPage Report</title>
    <link rel="stylesheet" href="style.css">
</head>
<body>
<main role="main">
    <span onclick="location.href='lsd-index.html';" title="back to index page" class="back_to_index"></span>
    <h1 class="logo">LSD</h1>
    <h1>Split by NewPage Report</h1>
    <div class="contents">
        <h3>Contents</h3>
        <ul class="contents">
            
            <li class="success"><a href="#4">A Success scenario</a></li>
            
        </ul>
    </div>
    

    <article id="4">
        <span onclick="location.href='#top';" title="top of page" class="back_to_top"></span>
        <h2 class="success">A Success scenario</h2>
        <section class="description success">
            <h3 class="success">Description</h3>
            Given a first scenario description<br/>When something happens<br/>Then something else happens
        </section>
        
        
        <section class="sequence diagram svg success">
            <h3 class="success">Sequence Diagram</h3>
            <figure><?xml version="1.0" encoding="UTF-8"?>

<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentStyleType="text/css" height="262px" preserveAspectRatio="none" style="width:109px;height:262px;background:#00000000;" version="1.1" viewBox="0 0 109 262" width="109px" zoomAndPan="magnify">
  <defs/>
  <g>
    <rect fill="none" height="36.7373" id="_title" style="stroke:none;stroke-width:0.0;" width="86" x="10" y="10"/>
    <text fill="#000000" font-family="Verdana" font-size="22" font-weight="bold" lengthAdjust="spacing" textLength="76" x="15" y="37.1182">A title</text>
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="57.75" x2="57.75" y1="130.752" y2="180.5513"/>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="36.75" y="127.8125">Arnie</text>
    <ellipse cx="58.25" cy="61.7373" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:1.0;"/>
    <path d="M58.25,69.7373 L58.25,96.7373 M45.25,77.7373 L71.25,77.7373 M58.25,96.7373 L45.25,111.7373 M58.25,96.7373 L71.25,111.7373 " fill="none" style="stroke:#000000;stroke-width:1.0;"/>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="36.75" y="193.6265">Arnie</text>
    <ellipse cx="58.25" cy="205.5659" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:1.0;"/>
    <path d="M58.25,213.5659 L58.25,240.5659 M45.25,221.5659 L71.25,221.5659 M58.25,240.5659 L45.25,255.5659 M58.25,240.5659 L71.25,255.5659 " fill="none" style="stroke:#000000;stroke-width:1.0;"/>
    <polygon fill="#000000" points="46.25,158.5513,56.25,162.5513,46.25,166.5513,50.25,162.5513" style="stroke:#000000;stroke-width:1.0;"/>
    <line style="stroke:#000000;stroke-width:1.0;" x1="22.25" x2="52.25" y1="162.5513" y2="162.5513"/>
    <a href="#1" target="_top" title="in" xlink:actuate="onRequest" xlink:href="#1" xlink:show="new" xlink:title="in" xlink:type="simple">
      <text fill="" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="12" x="29.25" y="157.8218">in</text>
    </a>
  </g>
</svg>
<?xml version="1.0" encoding="UTF-8"?>

<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentStyleType="text/css" height="278px" preserveAspectRatio="none" style="width:213px;height:278px;background:#00000000;" version="1.1" viewBox="0 0 213 278" width="213px" zoomAndPan="magnify">
  <defs/>
  <g>
    <rect fill="none" height="36.7373" id="_title" style="stroke:none;stroke-width:0.0;" width="190" x="10" y="10"/>
    <text fill="#000000" font-family="Verdana" font-size="22" font-weight="bold" lengthAdjust="spacing" textLength="180" x="15" y="37.1182">A second page</text>
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="100" x2="100" y1="132.7666" y2="182.5659"/>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="52" x="71" y="112.8125">Badboy</text>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="21" x="86.5" y="129.8271">DB</text>
    <path d="M82,62.7373 C82,52.7373 100,52.7373 100,52.7373 C100,52.7373 118,52.7373 118,62.7373 L118,88.7373 C118,98.7373 100,98.7373 100,98.7373 C100,98.7373 82,98.7373 82,88.7373 L82,62.7373 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.5;"/>
    <path d="M82,62.7373 C82,72.7373 100,72.7373 100,72.7373 C100,72.7373 118,72.7373 118,62.7373 " fill="none" style="stroke:#000000;stroke-width:1.5;"/>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="52" x="71" y="195.6411">Badboy</text>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="21" x="86.5" y="212.6558">DB</text>
    <path d="M82,225.5952 C82,215.5952 100,215.5952 100,215.5952 C100,215.5952 118,215.5952 118,225.5952 L118,251.5952 C118,261.5952 100,261.5952 100,261.5952 C100,261.5952 82,261.5952 82,251.5952 L82,225.5952 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.5;"/>
    <path d="M82,225.5952 C82,235.5952 100,235.5952 100,235.5952 C100,235.5952 118,235.5952 118,225.5952 " fill="none" style="stroke:#000000;stroke-width:1.5;"/>
    <polygon fill="#000000" points="133,160.5659,143,164.5659,133,168.5659,137,164.5659" style="stroke:#000000;stroke-width:1.0;"/>
    <line style="stroke:#000000;stroke-width:1.0;" x1="100" x2="139" y1="164.5659" y2="164.5659"/>
    <a href="#2" target="_top" title="out" xlink:actuate="onRequest" xlink:href="#2" xlink:show="new" xlink:title="out" xlink:type="simple">
      <text fill="" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="21" x="107" y="159.8364">out</text>
    </a>
  </g>
</svg>
<?xml version="1.0" encoding="UTF-8"?>

<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentStyleType="text/css" height="170px" preserveAspectRatio="none" style="width:431px;height:170px;background:#00000000;" version="1.1" viewBox="0 0 431 170" width="431px" zoomAndPan="magnify">
  <defs/>
  <g>
    <rect fill="none" height="36.7373" id="_title" style="stroke:none;stroke-width:0.0;" width="163" x="140.75" y="10"/>
    <text fill="#000000" font-family="Verdana" font-size="22" font-weight="bold" lengthAdjust="spacing" textLength="153" x="145.75" y="37.1182">A third page</text>
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="43" x2="43" y1="84.752" y2="134.5513"/>
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="390.5" x2="390.5" y1="84.752" y2="134.5513"/>
    <rect fill="#FFFFFF" height="31.0146" rx="2.5" ry="2.5" style="stroke:#000000;stroke-width:1.0;" width="46" x="20" y="52.7373"/>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="32" x="27" y="73.8125">Beta</text>
    <rect fill="#FFFFFF" height="31.0146" rx="2.5" ry="2.5" style="stroke:#000000;stroke-width:1.0;" width="46" x="20" y="133.5513"/>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="32" x="27" y="154.6265">Beta</text>
    <rect fill="#FFFFFF" height="31.0146" rx="2.5" ry="2.5" style="stroke:#000000;stroke-width:1.0;" width="69" x="356.5" y="52.7373"/>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="55" x="363.5" y="73.8125">Gamma</text>
    <rect fill="#FFFFFF" height="31.0146" rx="2.5" ry="2.5" style="stroke:#000000;stroke-width:1.0;" width="69" x="356.5" y="133.5513"/>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="55" x="363.5" y="154.6265">Gamma</text>
    <line style="stroke:#000000;stroke-width:2.0;" x1="373" x2="383" y1="111.5513" y2="121.5513"/>
    <line style="stroke:#000000;stroke-width:2.0;" x1="373" x2="383" y1="121.5513" y2="111.5513"/>
    <line style="stroke:#000000;stroke-width:1.0;" x1="43" x2="378" y1="116.5513" y2="116.5513"/>
    <a href="#3" target="_top" title="An interaction description that is long enough to need abbreviating" xlink:actuate="onRequest" xlink:href="#3" xlink:show="new" xlink:title="An interaction description that is long enough to need abbreviating" xlink:type="simple">
      <text fill="" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="324" x="50" y="111.8218">An interaction description that is long enough ...</text>
    </a>
  </g>
</svg>
</figure>
            <div onclick="location.href='#!';" id="5" class="overlay">
                <div class="popup" onclick="event.stopPropagation();">
                    <h2>PlantUml markup</h2>
                    <span class="copy" title="copy" onclick="copyToClipboard('sequence_uml_code5')"></span>
                    <a class="close" href="#!">&times;</a>
                    <div class="content">
                        <pre><code id="sequence_uml_code5">@startuml

!theme plain

skinparam backgroundColor transparent

hide unlinked


!include &lt;tupadr3/font-awesome-5/hamburger&gt;

!include &lt;tupadr3/font-awesome-5/heart&gt;



actor A as &quot;Arnie&quot;

boundary UnusedParticipant

database B as &quot;Badboy\nDB&quot;




title A title


?-&gt; A: &lt;text fill&#x3D;&quot;&quot;&gt;[[#1 {in} in]]&lt;/text&gt;


@enduml
@startuml

!theme plain

skinparam backgroundColor transparent

hide unlinked


!include &lt;tupadr3/font-awesome-5/hamburger&gt;

!include &lt;tupadr3/font-awesome-5/heart&gt;



actor A as &quot;Arnie&quot;

boundary UnusedParticipant

database B as &quot;Badboy\nDB&quot;




title A second page


B -&gt;?: &lt;text fill&#x3D;&quot;&quot;&gt;[[#2 {out} out]]&lt;/text&gt;


@enduml
@startuml

!theme plain

skinparam backgroundColor transparent

hide unlinked


!include &lt;tupadr3/font-awesome-5/hamburger&gt;

!include &lt;tupadr3/font-awesome-5/heart&gt;



actor A as &quot;Arnie&quot;

boundary UnusedParticipant

database B as &quot;Badboy\nDB&quot;




title A third page


Beta -&gt;x Gamma: &lt;text fill&#x3D;&quot;&quot;&gt;[[#3 {An interaction description that is long enough to need abbreviating} An interaction description that is long enough ...]]&lt;/text&gt;


@enduml</code></pre>
                    </div>
                </div>
            </div>
            <sub><a href="#5">[source]</a></sub>
            <sub><a href="#6">[component diagram]</a></sub>
        </section>
        <!-- Popup for component diagram -->
        <div onclick="location.href='#!';" id="6" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>Component Diagram</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre><?xml version="1.0" encoding="UTF-8"?>

<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentStyleType="text/css" height="192px" preserveAspectRatio="none" style="width:108px;height:192px;background:#00000000;" version="1.1" viewBox="0 0 108 192" width="108px" zoomAndPan="magnify">
  <defs/>
  <g>
    <g id="elem_Beta">
      <rect fill="#FFFFFF" height="56.7373" rx="2.5" ry="2.5" style="stroke:#000000;stroke-width:0.0;" width="90" x="12" y="9.7373"/>
      <rect fill="#FFFFFF" height="10" style="stroke:#000000;stroke-width:0.0;" width="15" x="82" y="14.7373"/>
      <rect fill="#FFFFFF" height="2" style="stroke:#000000;stroke-width:0.0;" width="4" x="80" y="16.7373"/>
      <rect fill="#FFFFFF" height="2" style="stroke:#000000;stroke-width:0.0;" width="4" x="80" y="20.7373"/>
      <text fill="#000000" font-family="Verdana" font-size="22" lengthAdjust="spacing" textLength="50" x="27" y="51.8555">Beta</text>
    </g>
    <g id="elem_Gamma">
      <ellipse cx="57" cy="110.7373" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:0.0;"/>
      <path d="M57,118.7373 L57,145.7373 M44,126.7373 L70,126.7373 M57,145.7373 L44,160.7373 M57,145.7373 L70,160.7373 " fill="none" style="stroke:#000000;stroke-width:0.0;"/>
      <text fill="#000000" font-family="Verdana" font-size="22" lengthAdjust="spacing" textLength="85" x="14.5" y="183.8555">Gamma</text>
    </g>
    <path d="M57,66.9562 C57,77.8232 57,90.5352 57,102.496 " fill="none" style="stroke:#000000;stroke-width:1.0;"/>
    <polygon fill="#000000" points="57,102.496,61,93.496,57,97.496,53,93.496,57,102.496" style="stroke:#000000;stroke-width:1.0;"/>
  </g>
</svg>
</pre>
                </div>
                <sub><a href="#uml6">[source]</a></sub>
            </div>
        </div>
        <!-- Popup for component diagram UML via component diagram popup -->
        <div onclick="location.href='#!';" id="uml6" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>PlantUml markup</h2>
                <span class="copy" title="copy" onclick="copyToClipboard('component_uml_code6')"></span>
                <a class="close" href="#6">&times;</a><!--show previous popup on close-->
                <div class="content">
                    <pre><code id="component_uml_code6">@startuml
!pragma layout smetana
!theme plain
skinparam backgroundColor transparent

actor A as &quot;Arnie&quot;
boundary UnusedParticipant
database B as &quot;Badboy\nDB&quot;


[Beta] --&gt; Gamma


remove @unlinked
@enduml</code></pre>
                </div>
            </div>
        </div>
        <!-- Section containing data for sequence diagram popups -->
        
        <div onclick="location.href='#!';" id="1" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>in</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>start some job</pre>
                </div>
            </div>
        </div>
        
        <div onclick="location.href='#!';" id="2" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>out</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>some data 1</pre>
                </div>
            </div>
        </div>
        
        <div onclick="location.href='#!';" id="3" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>An interaction description that is long enough ...</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>β</pre>
                </div>
            </div>
        </div>
        
        
    </article>
    
</main>
</body>
<footer>
    <script type = "text/javascript" src = "https://d3js.org/d3.v4.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.0.0/build/styles/default.min.css">
    <script src="custom.js"></script>
    <script language = "javascript">
        addSliderForSvgZoom();
        highlightLifelinesWhenClicked();
    </script>
    <script src="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.0.0/build/highlight.min.js"></script>
    <script>hljs.highlightAll();</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js"></script>
    <script>
        highlightKeywords();
        
            
        
        unMarkSvg();
    </script>
</footer>
</html>
